### PR TITLE
Changed InfiniteScroll to handle variable item heights

### PR DESCRIPTION
**Pull request by ericsoderberghp**
#### What does this PR do?

Changed InfiniteScroll to handle variable item heights.

It is an alternative approach to that taken in #4525. In that pull request, an average page height is used. In this one, we approximate un-rendered page heights but then once we render a page, we keep track of that page's height.

Also, this version handles browser resizing events as well.

#### Where should the reviewer start?

This is a fairly broad refactoring. The key state variable to keep an eye on is `pageHeights`.
Please flag anything that could be made clearer with more or better comments as well.

#### What testing has been done on this PR?

Storybook

#### How should this be manually tested?

Storybook

#### What are the relevant issues?

#4337 

#### Do the grommet docs need to be updated?

no

#### Should this PR be mentioned in the release notes?

yes

#### Is this change backwards compatible or is it a breaking change?

backwards compatible

### CODE REVIEW

**Comment by halocline** [comment]
👀

File: src/js/components/InfiniteScroll/InfiniteScroll.js [hunk 1]
```diff
@@ -9,196 +13,260 @@ import {
...
+    // might not be there yet or might have already rendered everything
+    if (belowMarkerRef.current) {
       scrollParents = findScrollParents(belowMarkerRef.current);
-      scrollParents.forEach(scrollParent =>
-        scrollParent.addEventListener('scroll', onScroll),
-      );
-      onScroll();
+      scrollParents.forEach(sp => sp.addEventListener('scroll', debounce));
     }
+    window.addEventListener('resize', debounce);
```
Comment (halocline) [inline src/js/components/InfiniteScroll/InfiniteScroll.js:125]: 👍

**Review by halocline — COMMENTED** [review]
Looking good. This is handling all the scenarios I have been looking at quite nicely.

I'd like to step away and come back to review the logic one more time to make sure I am following it correctly.

File: src/js/components/InfiniteScroll/InfiniteScroll.js [hunk 2]
```diff
@@ -9,196 +13,260 @@ import {
...
-        // clean up after having shown
-        setShow(undefined);
       }
     }, 100);
     return () => clearTimeout(timer);
-  }, [show]);
+  }, [renderPageBounds, show, step]);
+
+  // calculate and keep track of page heights
+  useLayoutEffect(() => {
```
[the remark this answers is not in the document]
Comment (ericsoderberghp) [reply src/js/components/InfiniteScroll/InfiniteScroll.js:177]: Exactly

File: src/js/components/InfiniteScroll/InfiniteScroll.js [hunk 3]
```diff
@@ -9,196 +13,260 @@ import {
...
+        }
+        i += 1;
+      }
+
+      if (nextPageHeights) setPageHeights(nextPageHeights);
+    }
+  }, [lastPage, pageHeights, renderPageBounds, replace, step]);
+
+  // calculate the height above the first rendered page using the pageHeights
+  const aboveHeight = useMemo(() => {
```
[the remark this answers is not in the document]
Comment (ericsoderberghp) [reply src/js/components/InfiniteScroll/InfiniteScroll.js:233]: good to hear! agreed

File: src/js/components/InfiniteScroll/InfiniteScroll.js (hunk 1 above)
Comment (ericsoderberghp) [reply src/js/components/InfiniteScroll/InfiniteScroll.js:125]: yep, icing on the cake

File: src/js/components/InfiniteScroll/InfiniteScroll.js [hunk 4]
```diff
@@ -209,73 +277,37 @@ const InfiniteScroll = ({
...
       />
     );
     if (renderMarker) {
       // need to give it a key
-      marker = React.cloneElement(renderMarker(marker), { key: 'below' });
+      const renderedMarker = renderMarker(marker);
+      marker = React.cloneElement(renderedMarker, {
+        key: 'below',
+        // https://github.com/facebook/react/issues/8873#issuecomment-489579878
+        ref: node => {
```
[the remark this answers is not in the document]
Comment (ericsoderberghp) [reply src/js/components/InfiniteScroll/InfiniteScroll.js:306]: Good question. I've added a comment in the code to attempt to explain it:
```
// We need to make sure our belowMarkerRef is tied to a component
// that has the same parent as the items being rendered. This is so
// we can use belowMarkerRef.current.parentNode.children to
// get a reference to the items in the DOM for calculating pageHeights.
```

**Review by halocline — APPROVED** [review]
Good to go by me. Nicely done.

**Review by ShimiSun — APPROVED** [review]

**Comment by ShimiSun** [comment]
🎉

**Pull request merged**